massa-execution/src/types.rs

```diff
@@ -26,6 +31,217 @@ pub(crate) struct StepHistoryItem {
 
     // list of SCE ledger changes caused by this execution step
     pub ledger_changes: SCELedgerChanges,
+
+    /// events produced during this step
+    pub events: EventStore,
+}
+
+/// Keep all events you need with some useful indexes
+#[derive(Default, Debug, Clone)]
+pub(crate) struct EventStore {
+    /// maps ids to events
+    id_to_event: Map<SCOutputEventId, SCOutputEvent>,
+
+    /// maps slot to a set of event ids
+    slot_to_id: HashMap<Slot, Set<SCOutputEventId>>,
+
+    /// maps initial caller to a set of event ids
+    caller_to_id: Map<Address, Set<SCOutputEventId>>,
+
+    /// maps direct event producer to a set of event ids
+    smart_contract_to_id: Map<Address, Set<SCOutputEventId>>,
+}
+
+impl EventStore {
+    /// add event to the store and all its indexes
+    pub fn insert(&mut self, id: SCOutputEventId, event: SCOutputEvent) {
+        if let Entry::Vacant(entry) = self.id_to_event.entry(id) {
+            self.slot_to_id
```

Comment:
I think you want to use [`entry`](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry), and in any case use something like [`or_insert_with`](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with) or [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) to only allocate the default when necessary(altough I guess if `default()` doesn't allocate anything these days it doesn't matter).

Reply:
changed